src/Interpreters/Cache/FileCacheFactory.cpp

```diff
@@ -47,20 +68,54 @@ FileCachePtr FileCacheFactory::create(const std::string & cache_name, const File
 
     auto cache = std::make_shared<FileCache>(cache_name, file_cache_settings);
     it = caches_by_name.emplace(
-        cache_name, std::make_unique<FileCacheData>(cache, file_cache_settings)).first;
+        cache_name, std::make_unique<FileCacheData>(cache, file_cache_settings, config_path)).first;
 
     return it->second->cache;
 }
 
-FileCacheFactory::FileCacheData FileCacheFactory::getByName(const std::string & cache_name)
+FileCacheFactory::FileCacheDataPtr FileCacheFactory::getByName(const std::string & cache_name)
 {
     std::lock_guard lock(mutex);
 
     auto it = caches_by_name.find(cache_name);
     if (it == caches_by_name.end())
         throw Exception(ErrorCodes::BAD_ARGUMENTS, "There is no cache by name: {}", cache_name);
 
-    return *it->second;
+    return it->second;
+}
+
+void FileCacheFactory::updateSettingsFromConfig(const Poco::Util::AbstractConfiguration & config)
+{
+    CacheByName caches_by_name_copy;
+    {
+        std::lock_guard lock(mutex);
+        caches_by_name_copy = caches_by_name;
+    }
+
+    for (const auto & [_, cache_info] : caches_by_name_copy)
+    {
+        if (cache_info->config_path.empty())
+            continue;
+
+        FileCacheSettings new_settings;
+        new_settings.loadFromConfig(config, cache_info->config_path);
+
+        FileCacheSettings old_settings;
+        {
+            std::lock_guard lock(cache_info->settings_mutex);
+            if (new_settings == cache_info->settings)
+                continue;
+
+            old_settings = cache_info->settings;
+        }
+
+        cache_info->cache->applySettingsIfPossible(new_settings, old_settings);
+
+        {
+            std::lock_guard lock(cache_info->settings_mutex);
+            cache_info->settings = old_settings;
```

Comment:
> Side question: what would be the source of truth, if some of the storage_conf.xml settings are ignored?

A query: `describe filesystem cache 'cache_name'`.

> If config is treated as a whole (all settings are either applied or rejected)

It is impossible to apply all the settings. Some settings can be applied only with server restart. But still there is a need sometimes to apply at least a subset of settings without restart as it can be done easily.

Reply:
> E.g. decreasing the number of threads while increasing the size of the individual queue looks like a valid use-case.

Yes, implemented support for decreasing the number of threads it in one of the last commits.